Sources/SourceKitLSP/Swift/SyntaxHighlightingTokenParser.swift

```diff
@@ -0,0 +1,245 @@
+         values.ref_function_constructor:
+      return (.method, [.static])
+    case values.ref_function_method_instance,
+         values.ref_function_destructor,
+         values.ref_function_subscript:
+      return (.method, [])
+    case values.decl_function_operator_prefix,
+         values.decl_function_operator_postfix,
+         values.decl_function_operator_infix:
+      return (.operator, [.declaration])
+    case values.ref_function_operator_prefix,
+         values.ref_function_operator_postfix,
+         values.ref_function_operator_infix:
+      return (.operator, [])
+    case values.decl_var_static,
+         values.decl_var_class,
+         values.decl_var_instance:
+      return (.property, [.declaration])
+    case values.decl_var_parameter:
+      // SourceKit seems to use these to refer to parameter labels,
+      // therefore we don't use .parameter here (which LSP clients like
+      // VSCode seem to interpret as variable identifiers, however
+      // causing a 'wrong highlighting' e.g. of `x` in `f(x y: Int) {}`)
+      return (.function, [.declaration])
+    case values.ref_var_static,
+         values.ref_var_class,
+         values.ref_var_instance:
+      return (.property, [])
+    case values.decl_var_local,
+         values.decl_var_global:
+      return (.variable, [.declaration])
+    case values.ref_var_local,
+         values.ref_var_global:
+      return (.variable, [])
+    case values.syntaxtype_comment,
+         values.syntaxtype_comment_marker,
+         values.syntaxtype_comment_url:
+      return (.comment, [])
+    case values.syntaxtype_doccomment,
+         values.syntaxtype_doccomment_field:
+      return (.comment, [.documentation])
+    case values.syntaxtype_type_identifier:
+      return (.type, [])
+    case values.syntaxtype_number:
+      return (.number, [])
+    case values.syntaxtype_string:
+      return (.string, [])
+    default:
+      let ignoredKinds: Set<sourcekitd_uid_t> = [
+        values.syntaxtype_identifier,
+        values.syntaxtype_string_interpolation_anchor,
+      ]
+      if !ignoredKinds.contains(uid) {
+        let name = api.uid_get_string_ptr(uid).map(String.init(cString:))
+        log("Unknown token kind: \(name ?? "?")", level: .debug)
+      }
+      return nil
+    }
+  }
+}
+
+extension Range where Bound == Position {
+  /// Splits a potentially multi-line range to multiple single-line ranges.
+  fileprivate func splitToSingleLineRanges(in snapshot: DocumentSnapshot) -> [Self] {
+    if isEmpty {
+      return []
+    }
+
+    if lowerBound.line == upperBound.line {
+      return [self]
+    }
+
+    guard let startIndex = snapshot.index(of: lowerBound),
+          let endIndex = snapshot.index(of: upperBound) else {
+      fatalError("Range \(self) reaches outside of the document")
+    }
+
+    let text = snapshot.text[startIndex..<endIndex]
+    let lines = text.split(separator: "\n", omittingEmptySubsequences: false)
+
+    return lines
```

Comment:
Would it make sense to use `lazy` here to avoid creating the intermediate arrays for `map`?